### PR TITLE
PHPC-941: Add toCanonicalJSON() and toRelaxedJSON() and update BSON corpus tests

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-941

### CODE REVIEW

File: scripts/convert-bson-corpus-tests.php [hunk 1]
```diff
@@ -2,6 +2,18 @@
...
+    'Double type: 1.23456789012345677E+18' => 'Variation in double\'s string representation (SPEC-850)',
+    'Double type: -1.23456789012345677E+18' => 'Variation in double\'s string representation (SPEC-850)',
+    'Int64 type: -1' => 'PHP encodes integers as 32-bit if range allows',
+    'Int64 type: 0' => 'PHP encodes integers as 32-bit if range allows',
+    'Int64 type: 1' => 'PHP encodes integers as 32-bit if range allows',
+    'Javascript Code with Scope: bad scope doc (field has bad string length)' => 'Depends on PHPC-889',
+    'Javascript Code with Scope: Unicode and embedded null in code string, empty scope' => 'Embedded null in code string is not supported in libbson (CDRIVER-1879)',
+    'Multiple types within the same document: All BSON types' => 'PHP encodes integers as 32-bit if range allows',
+    'Top-level document validity: Bad $date (number, not string or hash)' => 'Legacy extended JSON $date syntax uses numbers (CDRIVER-2223)',
+];
```
Comment (jmikola) [inline scripts/convert-bson-corpus-tests.php]: This is a welcome improvement inspired by [test-bson-corpus.c](https://github.com/mongodb/libbson/blob/4affa6aa91ce896eca7f1b4326b49bcf1d0c16f5/tests/test-bson-corpus.c#L19) in libbson. No longer will we need to manually insert (and restore) `--XFAIL--` blocks in generated test files.

File: src/BSON/functions.c [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+		return;
+	}
+
+	if (mode == PHONGO_JSON_MODE_LEGACY) {
+		json = bson_as_json(bson, &json_len);
+	} else if (mode == PHONGO_JSON_MODE_CANONICAL) {
+		json = bson_as_canonical_json(bson, &json_len);
+	} else if (mode == PHONGO_JSON_MODE_RELAXED) {
+		json = bson_as_relaxed_json(bson, &json_len);
+	}
```
Comment (jmikola) [inline src/BSON/functions.c]: The functions in this file were mostly copied as-is from `src/bson.c`. This is the only significant change, as I've added a mode enum and support for `bson_as_relaxed_json()`.

File: php_phongo.c [hunk 3]
```diff
@@ -2417,7 +2414,8 @@ static const zend_function_entry mongodb_functions[] = {
 	ZEND_NS_NAMED_FE("MongoDB\\BSON", fromPHP, PHP_FN(MongoDB_BSON_fromPHP), ai_bson_fromPHP)
 	ZEND_NS_NAMED_FE("MongoDB\\BSON", toPHP, PHP_FN(MongoDB_BSON_toPHP), ai_bson_toPHP)
 	ZEND_NS_NAMED_FE("MongoDB\\BSON", toJSON, PHP_FN(MongoDB_BSON_toJSON), ai_bson_toJSON)
-	ZEND_NS_NAMED_FE("MongoDB\\BSON", toExtendedJSON, PHP_FN(MongoDB_BSON_toExtendedJSON), ai_bson_toExtendedJSON)
+	ZEND_NS_NAMED_FE("MongoDB\\BSON", toCanonicalJSON, PHP_FN(MongoDB_BSON_toCanonicalJSON), ai_bson_toJSON)
+	ZEND_NS_NAMED_FE("MongoDB\\BSON", toRelaxedJSON, PHP_FN(MongoDB_BSON_toRelaxedJSON), ai_bson_toJSON)
```
Comment (derickr) [inline php_phongo.c]: Do I correctly understand that ``fromJSON`` will just read all of these?

File: src/BSON/functions.c [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+} /* }}} */
+
+static void phongo_bson_to_json(INTERNAL_FUNCTION_PARAMETERS, php_phongo_json_mode_t mode)
+{
+	char                *data;
+	phongo_zpp_char_len  data_len;
+	const bson_t        *bson;
+	bool                 eof = false;
+	bson_reader_t       *reader;
+	char                *json;
```
Comment (derickr) [inline src/BSON/functions.c]: Travis says about this:
```
/home/travis/build/mongodb/mongo-php-driver/src/BSON/functions.c: In function ‘phongo_bson_to_json.isra.2’:
/home/travis/.phpenv/versions/7.0.21/include/php/Zend/zend_string.h:159:8: error: ‘json’ may be used uninitialized in this function [-Werror=uninitialized]
/home/travis/build/mongodb/mongo-php-driver/src/BSON/functions.c:113:23: note: ‘json’ was declared here
```

And hence doesn't run the tests.

Comment (derickr) [inline src/BSON/functions.c]: It's logically not possible due to the way the enum works and how it is called, but the compiler doesn't know that.

File: php_phongo.c (hunk 3 above)
Comment (jmikola) [inline php_phongo.c]: Correct. The parser in libbson has logic to read all three formats. This was actually something I fixed in [CDRIVER-2223](https://jira.mongodb.org/browse/CDRIVER-2223).

**Pull request merged**